backbone.js

```diff
@@ -398,10 +401,20 @@
     this.attributes = {};
     if (options.collection) this.collection = options.collection;
     if (options.parse) attrs = this.parse(attrs, options) || {};
-    attrs = _.defaults({}, attrs, _.result(this, 'defaults'));
-    this.set(attrs, options);
-    this.changed = {};
-    this.initialize.apply(this, arguments);
+    var model = this, args = slice.call(arguments);
+    var defaults = _.result(this, 'defaults');
+    var consume = function(defaults) {
+      attrs = _.defaults({}, attrs, defaults);
+      model.set(attrs, options);
+      model.changed = {};
+      var init = model.initialize.apply(model, args);
+      if (isPromise(init))
+        return init.then(_.constant(model));
+      return model;
+    };
+    if (isPromise(defaults))
+      return defaults.then(consume);
+    return consume(defaults);
```

Comment:
Would love to hear those use cases if you feel like sharing.

On Mon, Jul 27, 2015, 9:06 AM Corin Lawson <notifications@github.com> wrote:

> In backbone.js
> <https://github.com/jashkenas/backbone/pull/3729#discussion_r35532739>:
>
> > -    this.changed = {};
> > -    this.initialize.apply(this, arguments);
> > +    var model = this, args = slice.call(arguments);
> > +    var defaults = _.result(this, 'defaults');
> > +    var consume = function(defaults) {
> > +      attrs = _.defaults({}, attrs, defaults);
> > +      model.set(attrs, options);
> > +      model.changed = {};
> > +      var init = model.initialize.apply(model, args);
> > +      if (isPromise(init))
> > +        return init.then(_.constant(model));
> > +      return model;
> > +    };
> > +    if (isPromise(defaults))
> > +      return defaults.then(consume);
> > +    return consume(defaults);
>
> I have real world use cases where initialize is not synchronous. Of
> course, the constructor will not return a Promise if the user never
> returns a Promise... but I can tell that you feel strongly about this and
> I'm happy to agree to disagree.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jashkenas/backbone/pull/3729/files#r35532739>.
>
-- 

Adam K (mobile)